### PR TITLE
MOB-23338: push token sync improvements

### PR DESCRIPTION
**Pull request by rymorale**
## Description
Update the push identifier sync network request to be sent only when the value changes. A new boolean configuration key value pair has been added `messaging.pushForceSync` which can be used to allow the push identifier sync request to be sent regardless if it has changed. As part of these improvements, there is also a 1 second timeout when sending a push identifier sync network request to prevent "spammy" behavior when the push identifier most likely hasn't changed since the last sync.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: TestApps/MessagingDemoAppSwiftUI/AppDelegate.swift [hunk 1]
```diff
@@ -55,10 +55,14 @@ final class AppDelegate: NSObject, UIApplicationDelegate {
...
             Messaging.updatePropositionsForSurfaces([cardSurface,cbeSurface1, cbeSurface2])
         }
-        
+        test()
         return true
     }
     
+    func test(blah: String = "blah") {
+        
+    }
```
Comment (kevinlind) [inline TestApps/MessagingDemoAppSwiftUI/AppDelegate.swift:64]: Is this still needed in the test app?

**Comment by kevinlind** [comment]
@rymorale Can you merge `main` branch in to `dev-v5.6.2` branch? That should clean up the fit-to-width changes in this PR which were already releases in v5.6.1.

File: AEPMessaging/Sources/MessagingProperties.swift [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+                _pushIdentifier = ServiceProvider.shared.namedKeyValueService.get(collectionName: MessagingConstants.DATA_STORE_NAME, key: MessagingConstants.NamedCollectionKeys.PUSH_IDENTIFIER) as? String
+            }
+            return _pushIdentifier
+        }
+
+        set {
+            /// Set the push identifier value to the local variable
+            _pushIdentifier = newValue
+            /// Save the new value to the named key-value service
+            ServiceProvider.shared.namedKeyValueService.set(collectionName: MessagingConstants.DATA_STORE_NAME, key: MessagingConstants.NamedCollectionKeys.PUSH_IDENTIFIER, value: newValue)
```
[the remark this answers is not in the document]
Comment (rymorale) [reply AEPMessaging/Sources/MessagingProperties.swift:31]: i'll get this fixed

File: AEPMessaging/Sources/MessagingProperties.swift [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        get {
+            /// Check if we have a push identifier set. if not, retrieve it from the named key-value service
+            if _pushIdentifier == nil {
+                _pushIdentifier = ServiceProvider.shared.namedKeyValueService.get(collectionName: MessagingConstants.DATA_STORE_NAME, key: MessagingConstants.NamedCollectionKeys.PUSH_IDENTIFIER) as? String
+            }
+            return _pushIdentifier
+        }
+
+        set {
+            if let _pushIdentifier = newValue, !_pushIdentifier.isEmpty {
```
Comment (kevinlind) [inline AEPMessaging/Sources/MessagingProperties.swift:28]: Looks like you are redefining `_pushIdentifier` locally here instead of updating the class `_pushIdentifier` variable. You'll need to add a line to update the class `_pushIdentifier` variable otherwise the variable will get out of sync with the persisted value.

File: AEPMessaging/Tests/UnitTests/MessagingPropertiesTests.swift [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+ the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ OF ANY KIND, either express or implied. See the License for the specific language
+ governing permissions and limitations under the License.
+ */
+
+import XCTest
+@testable import AEPMessaging
+import AEPServices
+
+class MessagingPropertiesTests: XCTestCase {
```
Comment (kevinlind) [inline AEPMessaging/Tests/UnitTests/MessagingPropertiesTests.swift:17]: Add a test which sets a pushIdentifier value, then sets a different value verifying the MessagingProperties local and persisted values are the new value.

File: AEPMessaging/Tests/UnitTests/MessagingTests.swift [hunk 5]
```diff
@@ -1079,6 +1083,114 @@ class MessagingTests: XCTestCase {
...
+        // test
+        XCTAssertNoThrow(messaging.handleProcessEvent(event))
+        XCTAssertNotNil(mockRuntime.dispatchedEvents)
+        let pushTokenEvent = mockRuntime.firstEvent
+        XCTAssertEqual(EventType.edge, pushTokenEvent?.type)
+        XCTAssertEqual(EventSource.requestContent, pushTokenEvent?.source)
+        XCTAssertEqual(NEW_PUSH_TOKEN, getDispatchedEventPushToken(event: pushTokenEvent))
+    }
+    
+    func testMultiplePushTokenSync_whenTokenMatchesAndWithinSyncTimeout_forceSyncTrue() {
```
Comment (kevinlind) [inline AEPMessaging/Tests/UnitTests/MessagingTests.swift:1171]: Add a test which checks a second setPushIdentifier request which matches the first does dispatch an Edge request if outside the timeout window. You do have a test above with pushForceSync = true, but that is comparing the stored and new token values where the last sync timestamp is nil.

File: AEPMessaging/Tests/UnitTests/MessagingTests.swift [hunk 6]
```diff
@@ -1191,6 +1191,48 @@ class MessagingTests: XCTestCase {
...
+    func testMultiplePushTokenSync_whenTokenMatchesAndOutsideSyncTimeout_forceSyncTrue() {
+        // setup
+        let delay = UInt32(1500000) // 1.5 seconds
+        let mockConfig = [EXPERIENCE_CLOUD_ORG: MOCK_EXP_ORG_ID, MessagingConstants.SharedState.Configuration.PUSH_FORCE_SYNC: true] as [String : Any]
+        let eventData: [String: Any] = [MessagingConstants.Event.Data.Key.PUSH_IDENTIFIER: MOCK_PUSH_TOKEN]
+
+        let event = Event(name: "handleProcessEvent", type: EventType.genericIdentity, source: EventSource.requestContent, data: eventData)
+        mockRuntime.simulateSharedState(for: MessagingConstants.SharedState.Configuration.NAME, data: (value: mockConfig, status: SharedStateStatus.set))
+        mockRuntime.simulateXDMSharedState(for: MessagingConstants.SharedState.EdgeIdentity.NAME, data: (value: SampleEdgeIdentityState, status: SharedStateStatus.set))
+        usleep(delay)
```
Comment (kevinlind) [inline AEPMessaging/Tests/UnitTests/MessagingTests.swift:1203]: nit: For these two sleeps, add a comment stating something to the effect of "sleeping to add delta between event timestamps".

**Review by kevinlind — APPROVED** [review]
